doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,349 @@
+they already maintain) to persistent storage. The resulting storage collections
+are exposed to the user as per-replica persist sources. As persistent storage
+is shared between all replicas, any replica is able to access all introspection
+sources and yield consistent introspection data back to the user.
+
+![Introspection data flow](static/introspection-data-flow.png)
+
+### COMPUTE
+
+#### Sinking Introspection Data
+
+Compute replicas already maintain introspection dataflows that feed arrangements
+that can be peeked by ADAPTER. These dataflows are extended to also sink their
+output to persistent storage.
+
+Eventually there will be a STORAGE client for interacting with storage
+collections, but it doesn't exist yet. Instead, replicas need to use the
+`persist` API directly. `persist` only abstracts a single storage shard that
+cannot be written to by multiple compute workers, so all introspection data from
+the same dataflow will be sinked through the same worker initially. We can still
+distribute the sinking of different introspection dataflows between different
+workers.
+
+The proposed design does not strictly require the existing arrangements, but we
+keep them in around for two reasons:
+
+  * Because persist compaction is not implemented yet, introspection storage
+    collections can grow without bounds. We therefore don't want to enable
+    sinking to persist for long-running Materialize deployments (like cloud). To
+    still be able to collect introspection data for such deployments, we need to
+    keep the existing mechanism of peeking introspection arrangements around
+    (even if it only works reliably for single-replica clusters).
+  * Once we have sinks working for replicated clusters, we'll want to think
+    about using them to sink the introspection arrangements instead of the
+    hand-rolled solution proposed in this design.
+
+To sink introspection data, replicas need the following storage connection
+information:
+
+  * Blob URI
+  * Consensus URI
+  * A shard ID for every log variant
+
+This connection information is passed to replicas through the
+`CreateInstance` compute command.
+
+#### Changes to the Compute API
+
+The `LoggingConfig` contained in `ComputeCommand::CreateInstance` is extended to
+also contain the storage connection information mentioned above (details subject
+to bike-shedding):
+
+```rust
+pub struct LoggingConfig {
+    pub granularity_ns: u128,
+    pub active_logs: HashMap<LogVariant, ActiveLog>,
+    pub log_logging: bool,
+    pub blob_uri: String,
+    pub consensus_uri: String,
+}
+
+pub struct ActiveLog {
+    pub index_id: GlobalId,
+    pub shard_id: ShardId,
+}
+```
+
+The compute controller generally assumes that it can send the same commands to
+all replicas in a cluster. This will not be the case anymore with the updated
+`LoggingConfig`, since shard IDs are specific to replicas. Therefore, the
+compute controller must take care of adjusting `CreateInstance` commands to the
```

Comment:
In principle, replicas could generate shard IDs themselves. They would still have to report those shard IDs back to controller/ADAPTER, since they are required for letting other replicas know about the introspection collections. But we could avoid replica-specific compute commands, I think.

The current design assumes that when a replica is restarted it continues writing to the same introspection shards/collections as before. Since replicas don't have persistent state, the shard IDs need to be stored in controller/ADAPTER and communicated to the replica on creation. So we can't allow the replica to generate its own shard IDs because we need to fix these values across restarts.

If instead a replica would start writing to new collections after each restart, then we could let the replicas generate the shard IDs themselves. However, this approach comes with a couple issues that I don't have a good solution for. See [this comment](https://github.com/MaterializeInc/materialize/pull/12399/files#r871013940).